In [3]:
import pandas as pd 
import numpy as np 
import matplotlib.pyplot as plt

# 1. Load the files 

In [4]:
recommendations = pd.read_csv("data/MSFT_recommendations.csv")
price = pd.read_csv("data/MSFT_price.csv")
income = pd.read_csv("data/MSFT_income.csv")
balance = pd.read_csv("data/MSFT_balance.csv")
cashflow = pd.read_csv("data/MSFT_cashflow.csv")
upgrades_downgrades = pd.read_csv("data/MSFT_upgrades_downgrades.csv")

# 2. Review the data

In [5]:
print(recommendations.head().T)

                                        0                              1  \
Date        1970-01-01 00:00:00.000000000  1970-01-01 00:00:00.000000001   
period                                 0m                            -1m   
strongBuy                              13                             12   
buy                                    40                             40   
hold                                    3                              3   
sell                                    0                              0   
strongSell                              0                              0   

                                        2                              3  
Date        1970-01-01 00:00:00.000000002  1970-01-01 00:00:00.000000003  
period                                -2m                            -3m  
strongBuy                               9                             10  
buy                                    42                             45  
hold            

In [6]:
print(price.head().T)

                    0           1           2           3           4
Date       1986-03-13  1986-03-14  1986-03-17  1986-03-18  1986-03-19
Adj_Close    0.059125    0.061237    0.062293    0.060709    0.059653
Close        0.097222    0.100694    0.102431    0.099826     0.09809
High         0.101563    0.102431    0.103299    0.103299    0.100694
Low          0.088542    0.097222    0.100694    0.098958    0.097222
Open         0.088542    0.097222    0.100694    0.102431    0.099826
Volume     1031788800   308160000   133171200    67766400    47894400


In [7]:
print(income.head().T)

                                                                 0  \
Date                                                    2025-06-30   
Tax_Effect_Of_Unusual_Items                            -77088000.0   
Tax_Rate_For_Calcs                                           0.176   
Normalized_EBITDA                                   160603000000.0   
Total_Unusual_Items                                   -438000000.0   
Total_Unusual_Items_Excluding_Goodwill                -438000000.0   
Net_Income_From_Continuing_Operation_Net_Minori...  101832000000.0   
Reconciled_Depreciation                              34153000000.0   
Reconciled_Cost_Of_Revenue                           87831000000.0   
EBITDA                                              160165000000.0   
EBIT                                                126012000000.0   
Net_Interest_Income                                    262000000.0   
Interest_Expense                                      2385000000.0   
Interest_Income     

In [8]:
print(balance.head().T)

                                                              0  \
Date                                                 2025-06-30   
Ordinary_Shares_Number                             7434158655.0   
Share_Issued                                       7434158655.0   
Net_Debt                                          12909000000.0   
Total_Debt                                        60588000000.0   
...                                                         ...   
Cash_Cash_Equivalents_And_Short_Term_Investments  94555000000.0   
Other_Short_Term_Investments                      64313000000.0   
Cash_And_Cash_Equivalents                         30242000000.0   
Cash_Equivalents                                  18531000000.0   
Cash_Financial                                    11711000000.0   

                                                              1  \
Date                                                 2024-06-30   
Ordinary_Shares_Number                             7434138859

In [9]:
print(cashflow.head().T)

                                                             0  \
Date                                                2025-06-30   
Free_Cash_Flow                                   71611000000.0   
Repurchase_Of_Capital_Stock                     -18420000000.0   
Repayment_Of_Debt                                -3216000000.0   
Issuance_Of_Debt                                           0.0   
Issuance_Of_Capital_Stock                         2056000000.0   
Capital_Expenditure                             -64551000000.0   
End_Cash_Position                                30242000000.0   
Beginning_Cash_Position                          18315000000.0   
Effect_Of_Exchange_Rate_Changes                     63000000.0   
Changes_In_Cash                                  11864000000.0   
Financing_Cash_Flow                             -51699000000.0   
Cash_Flow_From_Continuing_Financing_Activities  -51699000000.0   
Net_Other_Financing_Charges                      -2291000000.0   
Cash_Divid

In [10]:
print(upgrades_downgrades.head().T)

                                      0                    1  \
GradeDate           2026-06-04 14:19:27  2026-06-04 12:54:43   
Firm                           TD Cowen    Cantor Fitzgerald   
ToGrade                             Buy           Overweight   
FromGrade                           Buy           Overweight   
Action                             reit                 reit   
priceTargetAction             Maintains            Maintains   
currentPriceTarget                540.0                502.0   
priorPriceTarget                  540.0                502.0   

                                      2                    3  \
GradeDate           2026-06-04 12:44:28  2026-06-01 16:00:02   
Firm                           Citizens          Wells Fargo   
ToGrade               Market Outperform           Overweight   
FromGrade             Market Outperform           Overweight   
Action                             reit                 main   
priceTargetAction             Maintains

In [11]:
print(price.describe())

          Adj_Close         Close          High           Low          Open  \
count  10143.000000  10143.000000  10143.000000  10143.000000  10143.000000   
mean      69.989506     76.273102     77.017792     75.475608     76.263265   
std      119.464685    119.981232    121.046756    118.829505    119.985830   
min        0.054902      0.090278      0.092014      0.088542      0.088542   
25%        3.895945      6.406250      6.472656      6.316406      6.398438   
50%       19.524582     27.900000     28.200001     27.590000     27.870001   
75%       47.575237     55.205000     55.660000     54.464375     55.020000   
max      538.658508    542.070007    555.450012    540.770020    555.229980   

             Volume  
count  1.014300e+04  
mean   5.530477e+07  
std    3.792154e+07  
min    2.304000e+06  
25%    3.024135e+07  
50%    4.847840e+07  
75%    6.924320e+07  
max    1.031789e+09  


In [12]:
print(income.describe())

       Tax_Effect_Of_Unusual_Items  Tax_Rate_For_Calcs  Normalized_EBITDA  \
count                 4.000000e+00            4.000000       4.000000e+00   
mean                 -3.402550e+07            0.169750       1.248052e+11   
std                   6.637711e+07            0.026462       2.807279e+10   
min                  -9.991800e+07            0.131000       9.990500e+10   
25%                  -8.279550e+07            0.164750       1.038425e+11   
50%                  -3.996900e+07            0.179000       1.193565e+11   
75%                   8.801000e+06            0.184000       1.403192e+11   
max                   4.375400e+07            0.190000       1.606030e+11   

       Total_Unusual_Items  Total_Unusual_Items_Excluding_Goodwill  \
count         4.000000e+00                            4.000000e+00   
mean         -1.670000e+08                           -1.670000e+08   
std           4.055737e+08                            4.055737e+08   
min          -5.490000e+08

In [13]:
print(balance.describe())

       Ordinary_Shares_Number  Share_Issued      Net_Debt    Total_Debt  \
count            4.000000e+00  4.000000e+00  4.000000e+00  4.000000e+00   
mean             7.441074e+09  7.441074e+09  2.365175e+10  6.223750e+10   
std              1.531728e+07  1.531728e+07  1.266503e+10  3.302947e+09   
min              7.432000e+09  7.432000e+09  1.253300e+10  5.996500e+10   
25%              7.433604e+09  7.433604e+09  1.281500e+10  6.043225e+10   
50%              7.434149e+09  7.434149e+09  2.311200e+10  6.092900e+10   
75%              7.441619e+09  7.441619e+09  3.394875e+10  6.273425e+10   
max              7.464000e+09  7.464000e+09  3.585000e+10  6.712700e+10   

       Tangible_Book_Value  Invested_Capital  Working_Capital  \
count         4.000000e+00      4.000000e+00     4.000000e+00   
mean          1.349292e+11      2.941300e+11     5.976775e+10   
std           4.779864e+10      7.514216e+10     2.138525e+10   
min           8.772000e+10      2.163230e+11     3.444800e+10   

In [14]:
print(cashflow.describe())

       Free_Cash_Flow  Repurchase_Of_Capital_Stock  Repayment_Of_Debt  \
count    4.000000e+00                 4.000000e+00       4.000000e+00   
mean     6.757650e+10                -2.265375e+10      -1.101475e+10   
std      6.582370e+09                 7.026042e+09       1.237047e+10   
min      5.947500e+10                -3.269600e+10      -2.907000e+10   
25%      6.373050e+10                -2.485775e+10      -1.403475e+10   
50%      6.838000e+10                -2.033250e+10      -6.119500e+09   
75%      7.222600e+10                -1.812850e+10      -3.099500e+09   
max      7.407100e+10                -1.725400e+10      -2.750000e+09   

       Issuance_Of_Debt  Issuance_Of_Capital_Stock  Capital_Expenditure  \
count      4.000000e+00               4.000000e+00         4.000000e+00   
mean       6.098750e+09               1.941250e+09        -4.025525e+10   
std        1.219750e+10               1.041965e+08         1.847195e+10   
min        0.000000e+00               1.84

# 3. Data Cleaning & Preprocessing

In [15]:
print(income.isna().sum())

Date                                                          0
Tax_Effect_Of_Unusual_Items                                   0
Tax_Rate_For_Calcs                                            0
Normalized_EBITDA                                             0
Total_Unusual_Items                                           0
Total_Unusual_Items_Excluding_Goodwill                        0
Net_Income_From_Continuing_Operation_Net_Minority_Interest    0
Reconciled_Depreciation                                       0
Reconciled_Cost_Of_Revenue                                    0
EBITDA                                                        0
EBIT                                                          0
Net_Interest_Income                                           0
Interest_Expense                                              0
Interest_Income                                               0
Normalized_Income                                             0
Net_Income_From_Continuing_And_Discontin

* **Income Statement** — No missing values

In [16]:
print(balance.isna().sum())

Date                                                0
Ordinary_Shares_Number                              1
Share_Issued                                        1
Net_Debt                                            1
Total_Debt                                          1
                                                   ..
Cash_Cash_Equivalents_And_Short_Term_Investments    1
Other_Short_Term_Investments                        1
Cash_And_Cash_Equivalents                           1
Cash_Equivalents                                    1
Cash_Financial                                      1
Length: 79, dtype: int64


**Balance Sheet** - 1 missing row per column (2021 fiscal year incomplete)

print(cashflow.isna().sum())

* **Balance Sheet** — 1 missing row per column (2021 fiscal year incomplete)

In [17]:
print(recommendations.isna().sum())

Date          0
period        0
strongBuy     0
buy           0
hold          0
sell          0
strongSell    0
dtype: int64


* **Recommendations** — No missing values 

In [18]:
print(upgrades_downgrades.isna().sum())

GradeDate               0
Firm                    0
ToGrade                 1
FromGrade             468
Action                  0
priceTargetAction      41
currentPriceTarget      0
priorPriceTarget        0
dtype: int64


* there are 468 nulls in ***FROMGrade*** and they are expected and meaningful not a data error , because when analyst covers stock for the first time there is no previous grade

In [19]:
upgrades_downgrades["FromGrade"] = upgrades_downgrades["FromGrade"].fillna("New Coverage")

In [20]:
print(upgrades_downgrades["ToGrade"].value_counts())
print(upgrades_downgrades.shape)

ToGrade
Buy                  301
Outperform           245
Overweight           213
Neutral               53
Hold                  21
Market Perform        16
Sell                  13
Strong Buy            10
Underperform           6
Equal-Weight           6
Sector Outperform      5
Market Outperform      2
Underweight            2
Sector Perform         2
Sector Weight          1
Perform                1
Long-Term Buy          1
Name: count, dtype: int64
(899, 8)


In [21]:
upgrades_downgrades = upgrades_downgrades.dropna(subset=["ToGrade"])
upgrades_downgrades["GradeDate"] = pd.to_datetime(upgrades_downgrades["GradeDate"])

In [22]:
income["Date"] = pd.to_datetime(income["Date"])
balance["Date"] = pd.to_datetime(balance["Date"])
cashflow["Date"] = pd.to_datetime(cashflow["Date"])

In [23]:
income = income.dropna()
balance = balance.dropna()
cashflow = cashflow.dropna()

In [24]:
price["Date"] = pd.to_datetime(price["Date"])
price = price.sort_values("Date")

In [25]:
bullish = ["Buy", "Outperform", "Overweight", "Strong Buy", 
           "Sector Outperform", "Market Outperform", "Long-Term Buy"]

bearish = ["Sell", "Underperform", "Underweight"]

In [28]:
def map_recommendation (grade):
    if grade in bullish:
        return 2
    elif grade in bearish :
        return 0
    else:
        return 1

upgrades_downgrades["target"] = upgrades_downgrades["ToGrade"].apply(map_recommendation)
print(upgrades_downgrades["target"].value_counts())

target
2    777
1    100
0     21
Name: count, dtype: int64
